docs/long_term_roadmap.md

```diff
@@ -0,0 +1,242 @@
+community on a 1-2 year timeline.
+
+First, let's recap the goals of Torch-MLIR.
+
+Technically, the goal of Torch-MLIR is to bridge the PyTorch and MLIR
+ecosystems. That's vague, but it captures a very important property: Torch-MLIR
+is not in the business of "innovating" either on the frontend or backend sides.
+The project scope is to be an enabling connector between the two systems.
+
+Non-technically, Torch-MLIR's goal is not to be an end-to-end product, but a
+reliable piece of "off the shelf" infrastructure that system designers use as
+part of their larger end-to-end flows. The main users are expected to be
+"integrators", not end-users writing Python. This has the following facets:
+- Community: Users of Torch-MLIR should feel empowered to participate in the
+  community to get their questions resolved, or propose (and even implement)
+  changes needed for their use cases.
+- Ecosystem alignment: Users of Torch-MLIR should feel that the project is
+  aligned with all of the projects that it collaborates with, making it safe to
+  bet on for the long term.
+- Ease of use: Users of Torch-MLIR should feel that it "Just Works", or that
+  when it fails, it fails in a way that is easy to understand, debug, and fix.
+- Development: Torch-MLIR should be easy and convenient to develop.
+
+Today, much of the design space and the main problems have been identified, but
+larger-scale architectural and cross-project changes are needed to realize the
+right long-term design. This will allow us to reach a steady-state that best
+meets the goals above.
+
+## The main architectural changes
+
+As described in
+[architecture.md](architecture.md),
+Torch-MLIR can be split into two main parts: the "frontend" and the "backend".
+
+The main sources of brittleness, maintenance cost, and duplicated work across
+the ecosystem are:
+
+- The frontend work required to lower TorchScript to the backend contract.
+- The irregular support surface area of the large number of PyTorch ops across
+  the Linalg, TOSA, and MHLO backends.
+
+Most of this document describes long-term ecosystem changes that will address
+these, drastically improving Torch-MLIR's ability to meet its goals.
+
+## Roadmap
+
+### Refactoring the frontend
+
+The primary way to make the frontend more reliable is to leverage new PyTorch
+infrastructure that bridges from the PyTorch eager world into compiler-land.
+PyTorch has two main projects that together cover almost all user use cases and
+provide a technically sound, high quality-of-implementation path from user
+programs into the compiler.
+
+- [TorchDynamo](https://github.com/pytorch/torchdynamo) - TorchDynamo uses
+  tracing-JIT-like techniques and program slicing to extract traces of tensor
+  operations, which can then be passed to lower-level compilers. It works
+  seamlessly with unmodified user programs.
+- [FuncTorch](https://github.com/pytorch/functorch) - FuncTorch is basically JAX
```

Comment:
Thanks for the info! From the perspective of this diagram, we are mostly focusing on the ways that users interact with the compiler, the two being 1) Seamless with TorchDynamo or 2) JAX-like explicit capture with FuncTorch. Going into the specifics of mentioning AOTAutograd is probably one level too deep for this doc.

Reply:
@silvasean In the medium term, we expect that "explicit capture" will also sit behind Dynamo, just a mode where Dynamo will error (or bake in assumptions) instead of producing multiple graphs.